infra/blueprint-test/pkg/bpt/bpt.go

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package bpt defines a blueprint and implements default stages and execution order for a blueprint test
+
+package bpt
+
+import (
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/utils"
+	"github.com/mitchellh/go-testing-interface"
+	"github.com/stretchr/testify/assert"
+)
+
+// Blueprint represents a config that can be initialized, applied, verified and torndown.
+type Blueprint interface {
+	Init(*assert.Assertions)
+	Apply(*assert.Assertions)
+	Verify(*assert.Assertions)
+	Teardown(*assert.Assertions)
+}
+
+// BlueprintTest implements a generic blueprint
+type BlueprintTest struct {
+	init     func(*assert.Assertions) // custom init function
+	apply    func(*assert.Assertions) // custom apply function
+	verify   func(*assert.Assertions) // custom verify function
+	teardown func(*assert.Assertions) // custom teardown function
+	bp       Blueprint                // blueprint to be tested
+}
+
+type bptOption func(*BlueprintTest)
+
+func DefineInit(init func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.init = init
+	}
+}
+
+func DefineApply(apply func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.apply = apply
+	}
+}
+
+func DefineTeardown(teardown func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.teardown = teardown
+	}
+}
+
+func DefineVerify(verify func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.verify = verify
+	}
+}
+
+func (bt *BlueprintTest) Init(a *assert.Assertions) {
```

Comment:
I don't think this approach of checking `bt.bp` for the Init value at run-time is the right one, as it makes inspect the struct to understand what will happen challenging. Instead, when a new BlueprintTest is initialized we should default to setting the values to the DefaultInit for that blueprint.

Reply:
Switched to setting blueprint default fns at init